src/createQueryClient.ts

```diff
@@ -172,6 +173,104 @@ export function createQueryClient(options?: ClientOptions): QueryClient {
     return mutation
   }
 
+  const defineMutation: DefineMutation = <
+    TDefinedAction extends MutationAction,
+    TDefinedPlaceholder extends unknown,
+    TDefinedTags extends MutationTags<TDefinedAction>
+  >(action: TDefinedAction, definedOptions?: MutationOptions<TDefinedAction, TDefinedPlaceholder, TDefinedTags>) => {
+
+    const definedMutation: DefinedMutationFunction<TDefinedAction, TDefinedPlaceholder> = (parameters, options) => {
+      const mutation = definedUseMutation(options)
+
+      mutation.mutate(...parameters).catch(() => {
+        // silence here for now
+        // later we'll want to automatically revert the `setQueryDataBefore`
+      })
+
+      return mutation
+    }
+
+    const definedUseMutation: DefinedMutationComposition<TDefinedAction, TDefinedPlaceholder> = (options) => {
+      const { 
+        setQueryDataBefore: definedSetQueryDataBefore, 
+        setQueryDataAfter: definedSetQueryDataAfter, 
+        onExecute: definedOnExecute, 
+        onSuccess: definedOnSuccess,
+        onError: definedOnError
+      } = definedOptions ?? {}
+
+      const placeholder = options?.placeholder ?? definedOptions?.placeholder
+
+      const { 
+        setQueryDataBefore, 
+        setQueryDataAfter, 
+        onExecute, 
+        onSuccess,
+        onError
+      } = options ?? {}
+  
+      const mutation = createMutation(action, {
+        placeholder,
+        retries: options?.retries ?? definedOptions?.retries,
+        refreshQueryData: options?.refreshQueryData ?? definedOptions?.refreshQueryData,
+        tags: (data) => {
+          const definedTags = getAllTags(definedOptions?.tags, data)
+          const tags = getAllTags(options?.tags, data)
+
+          return [...definedTags, ...tags]
+        },
+        onExecute: (context) => {
+          if(setQueryDataBefore) {
+            const tags = getAllTags(options?.tags, undefined)
+            
+            setQueryData(tags, (queryData: QueryData): QueryData => setQueryDataBefore(queryData, context))
```

Comment:
this took me a second to figure out inline

```suggestion
            const tags = getAllTags(options?.tags, undefined)
            const setter = (queryData: QueryData): QueryData => setQueryDataBefore(queryData, context)
            
            setQueryData(tags, setter)
```

Reply:
this pattern exists in a few places, I'll let you decide if it's worth doing in the 3 other places

Reply:
TBH typescript is making this harder to read than it actually is... The js is pretty simple but typescript doesn't like the type of `queryData` without specifying it to something wider (because of the generics). 

Classic typescript... I'll break it out 👍 

```
setQueryData(tags, data => setQueryDataBefore(data, context))
```